test/e2e/suite/command/pull.go

```diff
@@ -437,6 +477,46 @@ var _ = Describe("OCI image layout users:", func() {
 			ORAS("pull", Flags.Layout, LayoutRef(root, multi_arch.Tag), "--platform", "linux/amd64", "-o", root).
 				MatchStatus(multi_arch.LinuxAMD64StateKeys, true, len(multi_arch.LinuxAMD64StateKeys)).Exec()
 		})
+
+		It("should preserve permissions", func() {
+			// prepare
+			pushRoot := GinkgoT().TempDir()
+			tag := "pushed_preserve_permissions"
+			ref := RegistryRef(ZOTHost, ArtifactRepo, tag)
+			Expect(CopyTestFiles(pushRoot)).ShouldNot(HaveOccurred())
+			ORAS("push", ref, foobar.PermissionDirectoryName).WithWorkDir(pushRoot).Exec()
+
+			// test
+			pullRoot := GinkgoT().TempDir()
+			ORAS("pull", "--preserve-permissions", ref, "-o", pullRoot).
+				WithDescription(pullRoot).Exec()
+
+			have, err := os.Stat(filepath.Join(pullRoot, foobar.PermissionFileName))
+			Expect(err).NotTo(HaveOccurred())
+			want, err := os.Stat(filepath.Join(pushRoot, foobar.PermissionFileName))
+			Expect(err).NotTo(HaveOccurred())
+			Expect(have.Mode().Perm()).To(Equal(want.Mode().Perm()))
+		})
+
+		It("should not preserve permissions", func() {
+			// prepare
+			pushRoot := GinkgoT().TempDir()
+			tag := "pushed_no_preserve_permissions"
+			ref := RegistryRef(ZOTHost, ArtifactRepo, tag)
+			Expect(CopyTestFiles(pushRoot)).ShouldNot(HaveOccurred())
+			ORAS("push", ref, foobar.PermissionDirectoryName).WithWorkDir(pushRoot).Exec()
+
+			// test
+			pullRoot := GinkgoT().TempDir()
+			ORAS("pull", ref, "-o", pullRoot).
+				WithDescription(pullRoot).Exec()
+
+			have, err := os.Stat(filepath.Join(pullRoot, foobar.PermissionFileName))
+			Expect(err).NotTo(HaveOccurred())
+			want, err := os.Stat(filepath.Join(pushRoot, foobar.NoPermissionFileName))
```

Comment:
It looks like we don't really need the `test-no-permission-file`? We can compare the same file in the pull root and push root like we did above, and their permissions should be different.